LoRaWAN: Remote Multicast Setup Service

martinjaeger opened the pull request:
This PR adds an implementation of Remote Multicast Setup ([TS005](https://resources.lora-alliance.org/technical-specifications/lorawan-remote-multicast-setup-specification-v1-0-0)) service specified by the LoRa Alliance.

Remote Multicast Setup is required for firmware-upgrade over the air (FUOTA). This PR is a continuation of PR #49284, adding only one service at a time to make review easier.

As for the rest of the LoRaWAN implementation in Zephyr, it uses some functions provided by the [LoRaMac-node module](https://github.com/zephyrproject-rtos/loramac-node) developed by Semtech.

The multicast process is started by the application and afterwards runs in the background in its own work queue thread.

**Testing**

The implementation has been successfully tested with AWS IoT.

martinjaeger committed:
lorawan: services: add Remote Multicast Setup

This service is responsible for multicast session key exchange and
setting up a class C session.

Signed-off-by: Martin Jäger <martin@libre.solar>

martinjaeger committed:
lorawan: services: add Class C session handling

The new functions are required for Multicast Class C session setup.

Signed-off-by: Martin Jäger <martin@libre.solar>

JordanYates commented on subsys/lorawan/services/lorawan_services.c:
```diff
@@ -160,6 +164,55 @@ int lorawan_services_reschedule_work(struct k_work_delayable *dwork, k_timeout_t
+{
+	int ret;
+
+	k_mutex_lock(&session_mutex, K_FOREVER);
```
I think this mutex setup can be achieved with the atomic API.
```
if (atomic_inc(&session_count) == 0) {
  lorawan_set_class(LORAWAN_CLASS_C);
  LOG_DBG("Switched to class C");
}
```
Are there realistic conditions under which `lorawan_set_class(LORAWAN_CLASS_C)` can fail?

martinjaeger replied:
Yes, it could fail when a transmission is active. Happened to me a couple of times during testing.

JordanYates commented on subsys/lorawan/services/lorawan_services.c:
```diff
@@ -160,6 +164,55 @@ int lorawan_services_reschedule_work(struct k_work_delayable *dwork, k_timeout_t
+
+	k_mutex_unlock(&session_mutex);
+
+	return ret;
```
If sticking with the mutex approach, this does not return the number of active sessions on the first call.

martinjaeger replied:
Good spot. Fixed.

JordanYates commented on subsys/lorawan/services/multicast.c:
```diff
@@ -0,0 +1,373 @@
+#include <zephyr/lorawan/lorawan.h>
+#include <zephyr/logging/log.h>
+#include <zephyr/random/random.h>
+#include <zephyr/settings/settings.h>
```
Persistent state management should be done through the `lorawan/lorawan_nvm.h` header to enable people who don't use the settings API to manage it through some other means.

martinjaeger replied:
I realized that the existing settings-based NVM backend already stores the multicast session details (hidden in `LORAMAC_NVM_NOTIFY_FLAG_MAC_GROUP2`), so I removed all the custom session data storage code.

JordanYates commented on subsys/lorawan/services/multicast.c:
```diff
@@ -0,0 +1,373 @@
+		uint8_t command_id = rx_buf[rx_pos++];
+
+		if (sizeof(tx_buf) - tx_pos < MAX_MULTICAST_ANS_LEN) {
+			LOG_ERR("insufficient tx_buf size, some requests discarded");
```
Under what conditions can this occur?

martinjaeger replied:
In theory we can have any number of batch requests inside one LoRaWAN message (up to the size of the LoRaWAN payload). However, I have never observed that in real life (using AWS and ChirpStack backends). Added a comment to make it a bit more clear.

JordanYates commented on subsys/lorawan/services/multicast.c:
```diff
@@ -0,0 +1,373 @@
+		case MULTICAST_CMD_MC_GROUP_SETUP: {
+			uint8_t id = rx_buf[rx_pos++] & 0x03;
+
+			ctx[id].mc_group.addr = rx_buf[rx_pos++];
```
These would be clearer as the following IMO:
```
ctx[id].mc_group.addr = sys_get_le32(rx_buf + rx_pos);
rx_pos += sizeof(uint32_t)
```
And a `memcpy` for the key.

martinjaeger replied:
Agreed and fixed.

JordanYates commented on subsys/lorawan/services/multicast.c:
```diff
@@ -0,0 +1,373 @@
+	return LoRaMacMcChannelSetup(&channel);
+}
+
+static void multicast_package_callback(uint8_t port, bool data_pending, int16_t rssi, int8_t snr,
```
Can this function be split up a bit?
The nesting starts getting extreme down in `MULTICAST_CMD_MC_CLASS_C_SESSION` handling.

martinjaeger replied:
Yep, makes sense. Done.

JordanYates reviewed (changes requested):
General implementation comments without testing

martinjaeger committed:
samples: subsys: lorawan: class_a: add multicast support

This is mainly to make sure that the multicast code is built in CI.

Signed-off-by: Martin Jäger <martin@libre.solar>

martinjaeger commented:
Thanks @JordanYates for the quick and very valuable review. I have now updated the PR based on your comments and re-tested everything with AWS IoT.

Also added an overlay config to the sample to make sure the multicast setup is built in CI. This will also be ensured in the future by the FUOTA sample, but as long as that's not there we can use the existing sample for build-testing.

martinjaeger commented:
@JordanYates would you have time for another look at the updates based on your review?

If possible, I'd like to get this in before feature freeze for v3.6 (Feb 2nd).

JordanYates commented on include/zephyr/lorawan/lorawan.h:
```diff
@@ -371,6 +371,20 @@ int lorawan_clock_sync_get(uint32_t *gps_time);
+ *
+ * @return 0 if successful, negative errno otherwise.
+ */
+int lorawan_remote_multicast_run(void);
```
Why can't this be run automatically in `SYS_INIT(POST_KERNEL, 1)`? From what I can tell it is only initialising memory, and the downlink list init is run in `SYS_INIT(POST_KERNEL, 0)`.

martinjaeger replied:
The idea was that the LoRaWAN services should only be run after the network was joined successfully, so the setup/starting of the service must be controlled by the application. As far as I know, we don't have a LoRaWAN API to determine if the device has successfully joined a network apart from the return code of the `lorawan_join` call.

JordanYates replied:
But that function doesn't schedule any transmissions or do anything that relies upon being connected.
It registers for downlink callbacks, and initialises two `k_work` structures.

martinjaeger replied:
Yeah, it would still react on incoming frames and do things like switching to class C. I thought it would be good to leave it to the application to enable the function. But it's probably fair to assume that when you compile multicast setup in you also definitely want to use it.

Updated to use `SYS_INIT` now.

JordanYates commented on subsys/lorawan/services/multicast.c:
```diff
@@ -0,0 +1,279 @@
+static void multicast_package_callback(uint8_t port, bool data_pending, int16_t rssi, int8_t snr,
+				       uint8_t len, const uint8_t *rx_buf)
+{
+	uint8_t tx_buf[3 * MAX_MULTICAST_ANS_LEN];
```
Where does the 3 come from here?
Is it just the maximum number of requests you've observed in a single packet?

martinjaeger replied:
It was just a random number to reserve a bit more space than just for one message. The standard states "A message MAY carry more than one command.", but there is no limit specified, which is a bit annoying.

I have now created a define for that and explained the decision in a comment.

We may want to turn this into a Kconfig if you prefer, even though I don't think anyone will ever need more than 3 messages.

JordanYates commented on subsys/lorawan/services/multicast.c:
```diff
@@ -0,0 +1,279 @@
+/* maximum length of multicast answers */
+#define MAX_MULTICAST_ANS_LEN 5
+
+#define MULTICAST_SETTINGS_BASE "lorawan/multicast"
```
I don't think this is needed anymore

martinjaeger replied:
Good spot, removed.

JordanYates commented on subsys/lorawan/services/Kconfig:
```diff
@@ -54,4 +54,14 @@ config LORAWAN_APP_CLOCK_SYNC_PERIODICITY
+	  Enables the LoRaWAN Remote Multicast Setup service according to
+	  TS005-1.0.0 as published by the LoRa Alliance.
+
+	  The service uses the default port 200.
```
Uses the default multicast port, not the default port

martinjaeger replied:
This would be confusing, as one could think the multicast messages are also sent on this port, which is not the case.

This Kconfig enables the LoRaWAN Remote Multicast Setup service and uses the default port 200 for this service. I think it's clear as it is. And the wording is exactly the same as for the already existing Application Layer Clock Sync service, btw.

JordanYates replied:
200 is not the "default port" any more than 202 is the "default port" because of `LORAWAN_PORT_CLOCK_SYNC`.
It's the default multicast port. I agree with the first sentence, so maybe "default downlink multicast port".

This can definitely be implied from the context that its in, but being explicit is rarely bad

martinjaeger replied:
I find "default downlink multicast port" even more confusing. Multicast downlinks can be sent using any port. The FUOTA fragments are all sent as multicast downlink messages, but on port 201, which is the default for fragmented data. The multicast setup service which we are talking about actually only sends singlecast messages on that port 200.

We could say "The service uses port 200, which is the default port for the Remote Multicast Setup service.". But that repeats almost the entire sentence above.

I prefer keeping it as is.

martinjaeger replied:
Cool, I agree with that. What about renaming to `LORAWAN_PORT_MULTICAST_SETUP`?

martinjaeger replied:
Done.

gdb3-ampaworks commented:
Excited to see this being added! I am working on an IoT product using LoRa and would like the FUOTA service.

I'll check out your branches/other PRs and see if I can test them out.

martinjaeger commented:
> I'll check out your branches/other PRs and see if I can test them out.

Sounds great. The latest branch with the complete FUOTA tested and working is [this one](https://github.com/martinjaeger/zephyr/tree/lorawan-fuota-jiapengli_20240124). I'll continue with the frag decoder PR once the multicast is merged.

JordanYates commented on samples/subsys/lorawan/class_a/overlay-multicast.conf:
```diff
@@ -0,0 +1,9 @@
+# NVS required to store LoRaWAN DevNonce and multicast sessions
+CONFIG_NVS=y
+CONFIG_SETTINGS=y
+CONFIG_LORAWAN_NVM_SETTINGS=y
```
Is this required? The multicast code doesn't do anything extra with settings from what I can see.

martinjaeger replied:
Yes, it is required. The multicast keys are stored in NVM. Backends expect multicast sessions to be persistent once set up.

This is also the reason why multicast depends on `LORAWAN_NVM_SETTINGS`.

JordanYates replied:
Expecting it to be persistent is different from requiring `LORAWAN_NVM_SETTINGS`. For example I use a NVS based backend for LoRaWAN state. I imagine that would work fine, but is not accurately expressed by `depends on LORAWAN_NVM_SETTINGS`. I think `depends on !LORAWAN_NVM_NONE` is more accurate?

martinjaeger replied:
That means you have added a custom backend that would also store *all* settings from the LoRaMAC-node stack and not only the DevNonce? I can switch the dependency to `!LORAWAN_NVM_NONE`, but for the in-tree sample we still need the settings, right?

martinjaeger replied:
Updated Kconfig dependencies to:

```
	depends on LORAWAN_NVM
	depends on !LORAWAN_NVM_NONE
```

JordanYates commented on samples/subsys/lorawan/class_a/sample.yaml:
```diff
@@ -49,3 +49,8 @@ tests:
     filter: CONFIG_ENTROPY_HAS_DRIVER
     integration_platforms:
       - nucleo_wl55jc
+  sample.lorawan.class_a.multicast:
```
This is no longer `class_a` only. I think it would make more sense as a separate sample.
This can be done in a later PR if you would prefer.

martinjaeger replied:
I have a dedicated FUOTA sample ready, which will be added with the last step, the fragmented data transport service. In that step we can remove these bits from the class_a sample again, but we need this for now to make it compile in CI.
